### PR TITLE
WebXRManager: Use async/await in setSession().

### LINKED ISSUE #20715: WebGL context must be marked as XR compatible in order to use with an immersive XRSession
When calling `WebXRManager.setSession( value )` you sometimes might get this message:

`Uncaught (in promise) DOMException: Failed to construct 'XRWebGLLayer': WebGL context must be marked as XR compatible in order to use with an immersive XRSession`

**To Reproduce**

I am able to reproduce the issue only on my localhost environment. When serving my viewer from the server it's working fine.

The reason for this issue is a racing condition.

when creating an instance of `XRWebGLLayer` the WebGL context `(gl)` should already be XR compatible enabled.

The `gl.makeXRCompatible()` function returns a `Promise`.

***Current setSession Code***

```js
	this.setSession = function ( value ) {

		session = value;

		if ( session !== null ) {

			session.addEventListener( 'select', onSessionEvent );
			session.addEventListener( 'selectstart', onSessionEvent );
			session.addEventListener( 'selectend', onSessionEvent );
			session.addEventListener( 'squeeze', onSessionEvent );
			session.addEventListener( 'squeezestart', onSessionEvent );
			session.addEventListener( 'squeezeend', onSessionEvent );
			session.addEventListener( 'end', onSessionEnd );

			const attributes = gl.getContextAttributes();

			if ( attributes.xrCompatible !== true ) {

				gl.makeXRCompatible();

			}

			const layerInit = {
				antialias: attributes.antialias,
				alpha: attributes.alpha,
				depth: attributes.depth,
				stencil: attributes.stencil,
				framebufferScaleFactor: framebufferScaleFactor
			};

			// eslint-disable-next-line no-undef
			const baseLayer = new XRWebGLLayer( session, gl, layerInit );

			session.updateRenderState( { baseLayer: baseLayer } );

			session.requestReferenceSpace( referenceSpaceType ).then( onRequestReferenceSpace );

			//

			session.addEventListener( 'inputsourceschange', updateInputSources );

		}

	};
```

***Suggested setSession Code***
```js
	this.setSession = function ( value ) {

		session = value;

		if ( session !== null ) {

			session.addEventListener( 'select', onSessionEvent );
			session.addEventListener( 'selectstart', onSessionEvent );
			session.addEventListener( 'selectend', onSessionEvent );
			session.addEventListener( 'squeeze', onSessionEvent );
			session.addEventListener( 'squeezestart', onSessionEvent );
			session.addEventListener( 'squeezeend', onSessionEvent );
			session.addEventListener( 'end', onSessionEnd );

			const attributes = gl.getContextAttributes();

			const setSession = function() {
				const layerInit = {
					antialias: attributes.antialias,
					alpha: attributes.alpha,
					depth: attributes.depth,
					stencil: attributes.stencil,
					framebufferScaleFactor: framebufferScaleFactor
				};
	
				// eslint-disable-next-line no-undef
				const baseLayer = new XRWebGLLayer( session, gl, layerInit );
	
				session.updateRenderState( { baseLayer: baseLayer } );
	
				session.requestReferenceSpace( referenceSpaceType ).then( onRequestReferenceSpace );
	
				//
	
				session.addEventListener( 'inputsourceschange', updateInputSources );
			};

			if ( attributes.xrCompatible !== true ) {

				gl.makeXRCompatible( setSession );

			} else {

				setSession();

			}

		}

	};
```

**Screenshots**
image

**Platform:**

 - Device: Mobile (Samsung Galaxy S20 Ultra - Exynos variant)
 - OS: Android 11
 - Browser: Chrome Canary 89.0.4329.2
 - Three.js version: dev, r122

I can create a pull request myself, I just remember that the way to do is on this repo isn't so straghtfarward.

Fork dev?

Build and then push?

If someone wants to do it be my guest. If not, please provide a link or explains how to do it right.

Thanks.

### PR DESCRIPTION
**Pull request by Mugen87**
Related issue: Fixed #20715.

**Description**

@ranbuch Do you mind testing this change? I've tried it with an Oculus Quest and it seems to work.

### CODE REVIEW

File: src/renderers/webxr/WebXRManager.js [hunk 1]
```diff
@@ -190,12 +177,13 @@ function WebXRManager( renderer, gl ) {
 			session.addEventListener( 'squeezestart', onSessionEvent );
 			session.addEventListener( 'squeezeend', onSessionEvent );
 			session.addEventListener( 'end', onSessionEnd );
+			session.addEventListener( 'inputsourceschange', onInputSourceChange );
```
Comment (Mugen87) [inline src/renderers/webxr/WebXRManager.js:180]: Was there a reason for not adding the `inputsourceschange` event listener at this place?

File: src/renderers/webxr/WebXRManager.js [hunk 2]
```diff
@@ -190,12 +177,13 @@ function WebXRManager( renderer, gl ) {
 			session.addEventListener( 'squeezestart', onSessionEvent );
 			session.addEventListener( 'squeezeend', onSessionEvent );
 			session.addEventListener( 'end', onSessionEnd );
+			session.addEventListener( 'inputsourceschange', onInputSourcesChange );
```
Comment (Mugen87) [inline src/renderers/webxr/WebXRManager.js:180]: Was there a reason for not adding the `inputsourceschange` event listener at this place?

**Comment by Mugen87** [comment]
Oh, it seems `misc_legacy` is failing. The build `three.js` produces a strange runtime error now:

> three.js:16787 Uncaught ReferenceError: regeneratorRuntime is not defined

Looks like `babel`  has issues to transpile the code...

**Comment by marcofugaro** [comment]
@Mugen87 I'll look into it!

**Comment by cabanier** [comment]
@mrdoob Without this change, three.js experiences will start failing as soon as Chrome 88 ships.
Can we merge it?

**Comment by Mugen87** [comment]
@mrdoob We also need this #20761 otherwise the build will fail.

**Comment by cabanier** [comment]
@Mugen87 it seems this PR is still failing. Are you looking into fixing it?

**Comment by Mugen87** [comment]
> Are you looking into fixing it?

Yes. I'm on it 😊 .

**Comment by Mugen87** [comment]
Okay, the problem is that `misc_legacy` is failing because the PR #20761 is missing. 

It is required so babel can correctly transpile the code for the build files `three.js` and `three.min.js`.

**Comment by toji** [comment]
Just voicing my concern as well that this is able to land quickly. Happy to contribute directly if I can, though I'm not familiar with the blocking dependency.

**Comment by mrdoob** [comment]
Thanks!

**Comment by toji** [comment]
Thank you!

**Comment by jpryne** [comment]
Just upgraded to Chrome 88.0.4324.96 & I'm facing this issue starting a WebXR session. Windows 10 64, Oculus Rift CV1

**Pull request merged**